packages/drivers/odsp-driver/src/fetchWithEpochValidation.ts

```diff
@@ -0,0 +1,125 @@
+    constructor(
+        private readonly persistedCache: IPersistedCache,
+        private readonly logger?: ITelemetryLogger) {
+    }
+
+    public set hashedDocumentId(docId: string | undefined) {
+        this._hashedDocumentId = docId;
+    }
+
+    public get hashedDocumentId(): string | undefined {
+        return this._hashedDocumentId;
+    }
+
+    public get fluidEpoch() {
+        return this._fluidEpoch;
+    }
+
+    public async fetchFromCache<T>(entry: ICacheEntry, maxOpCount: number | undefined): Promise<T | undefined> {
+        const value: IPersistedCacheValue = await this.persistedCache.get(entry, maxOpCount);
+        if (value !== undefined) {
+            try {
+                this.validateEpochFromResponse(value.fluidEpoch);
+            } catch (error) {
+                this.checkForEpochError(createOdspNetworkError("Epoch Mismatch", fluidEpochMismatchError));
+            }
+            return value.value as T;
+        }
+        return undefined;
+    }
+
+    public async fetchAndParseAsJSON<T>(
+        url: string,
+        fetchOptions: {[index: string]: any},
+        addInBody: boolean = false,
+    ): Promise<IOdspResponse<T>> {
+        // Add epoch either in header or in body.
+        this.addEpochInRequest(fetchOptions, addInBody);
+        try {
+            const response = await fetchAndParseAsJSONHelper<T>(url, fetchOptions);
+            this.validateEpochFromResponse(response.headers.get("x-fluid-epoch"));
+            return response;
+        } catch (error) {
+            this.checkForEpochError(error);
+            throw error;
+        }
+    }
+
+    public async fetchResponse(
+        url: string,
+        fetchOptions: {[index: string]: any},
+        addInBody: boolean = false,
+    ): Promise<Response> {
+        // Add epoch either in header or in body.
+        this.addEpochInRequest(fetchOptions, addInBody);
+        try {
+            const response = await fetchHelper(url, fetchOptions);
+            this.validateEpochFromResponse(response.headers.get("x-fluid-epoch"));
+            return response;
+        } catch (error) {
+            this.checkForEpochError(error);
+            throw error;
+        }
+    }
+
+    private addEpochInRequest(fetchOptions: {[index: string]: any}, addInBody: boolean) {
+        if (this.fluidEpoch !== undefined) {
+            if (addInBody) {
+                // We use multi part form request for post body where we want to use this.
+                // So extract the form boundary to mark the end of form.
+                let body: string = fetchOptions.body;
+                const formBoundary = body.split("\r\n")[0].substring(2);
+                body += `\r\nepoch=${this.fluidEpoch}\r\n`;
+                body += `\r\n--${formBoundary}--`;
+                fetchOptions.body = body;
+                return;
+            }
+            // Else add in headers.
+            fetchOptions.headers = {
+                ...fetchOptions.headers,
```

Comment:
This will result in extra OPTIONS call. We should talk to SPO folks to see if we either use POST everywhere, or can pass it as query param

Reply:
So POST call is not enabled for all calls. It is just enabled for trees latest for now, but they are working to enable it for all apis. The argument to add it in body is to use the new POST api. So if the caller wants to add it in body, then they will pass the addInBody as true. 
Otherwise now I am adding it to query param, if the length is less than 2048 after adding otherwise to headers.